qiskit/pulse/library/parametric_pulses.py

```diff
@@ -37,68 +37,224 @@ class ParametricPulseShapes(Enum):
+    Then, each output sample :math:`y` is modified according to:
+
+    .. math::
+
+        y \mapsto A\frac{y-y^*}{A-y^*},
+
+    where :math:`y^*` is the value of the un-normalized Gaussian .
+    This sets the endpoints to :math:`0` while preserving the amplitude at the center.
+    If :math:`A=y^*`, :math:`y` is set to :math:`1`.
+    The endpoints are at ``x = -1, x = duration + 1``.
+
+    Integrated area under the full curve is ``amp * np.sqrt(2*np.pi*sigma**2)``
+
+    Args:
+        t: Symbol object represents time.
+        center: Symbol or expression represents the middle point of the samples.
+        zeroed_width: Symbol or expression represents the endpoints the samples.
+        sigma: Symbol or expression represents Gaussian sigma.
+
+    Returns:
+        Symbolic equation.
+    """
+    gauss = sym.exp(-((t - center) / sigma) ** 2 / 2)
+
+    t_edge = zeroed_width / 2
+    offset = sym.exp(-(t_edge / sigma) ** 2 / 2)
+
+    return (gauss - offset) / (1 - offset)
 
 
 class ParametricPulse(Pulse):
     """The abstract superclass for parametric pulses."""
+    __slots__ = ("param_values", "type")
+
+    PARAM_DEF = ["duration"]
+
+    definition = None
+    constraints = None
+
+    def __init_subclass__(cls, **kwargs):
+        # caching lambda symbolic equation for better performance
+
+        # Lambdify the definition
+        if cls._define():
+            params = ["t"] + cls.PARAM_DEF
+            cls.definition = staticmethod(lambdify_symbolic_pulse(cls._define(), params))
+        else:
+            raise PulseError(f"'{cls.__name__}' class has no waveform definition.")
+
+        # Lambdify the constraints
+        if cls._constraints():
+            params = cls.PARAM_DEF + ["limit"]
+            cls.constraints = [lambdify_symbolic_pulse(c, params) for c in cls._constraints()]
+        else:
+            cls.constraints = sym.true
 
     @abstractmethod
     def __init__(
         self,
         duration: Union[int, ParameterExpression],
+        parameters: Optional[Tuple[Union[ParameterExpression, complex]]] = None,
         name: Optional[str] = None,
         limit_amplitude: Optional[bool] = None,
+        type: Optional[str] = None,
     ):
         """Create a parametric pulse and validate the input parameters.
 
         Args:
-            duration: Pulse length in terms of the the sampling period `dt`.
+            duration: Pulse length in terms of the sampling period `dt`.
+            parameters: Other parameters to form waveform.
             name: Display name for this pulse envelope.
             limit_amplitude: If ``True``, then limit the amplitude of the
-                             waveform to 1. The default is ``True`` and the
-                             amplitude is constrained to 1.
+                waveform to 1. The default is ``True`` and the
+                amplitude is constrained to 1.
+            type: Type of this waveform. This appears in the representation string.
```

Comment:
Descriptor would be the best solution https://github.com/Qiskit/qiskit-terra/pull/7821/commits/d763f48ed34f492a2ac70c3979473049c5048598. We can QPY serialize the state of descriptor since it can store definitions of all pulse types. Then we can regenerate custom Types before loading actual pulse instructions. Likely this will simplify the framework for loading custom pulses.